feat: `Role`, `ClusterRole`, `RoleBinding`, `ClusterRoleBinding` L2s

Chriscbr opened the pull request:
Fixes #24 
Fixes #25
Fixes #26
Fixes #374 

Adds Role and ClusterRole L2s with "allowXxx" methods that resemble those available in the AWS CDK. This allows you to write code like:

```ts
import * as kplus from 'cdk8s-plus-23';

declare const deployment: kplus.Deployment;

const role = new kplus.ClusterRole(chart, 'my-cluster-role');
role.allowRead(deployment, kplus.ApiResource.SECRETS, kplus.ApiResource.PODS);
```

...where the "read" syntactic sugar grants permissions to the "get", "list", and "watch" commands. There are other allow commands available that offer more granularity if needed.

RoleBinding and ClusterRoleBinding L2s have also been created, and they can be instantiated through their constructors, or created using the appropriate bind methods that will attach it to the role automatically:

```ts
declare const role: kplus.Role;
declare const clusterRole: kplus.ClusterRole;

const user = new kplus.User({ name: 'alice@example.com' });
const group = new kplus.Group({ name: 'frontend-devs' });

role.bind(user, group); // creates a RoleBinding associated with the same namespace `role` was defined with
clusterRole.bindInNamespace('development', user, group); // creates a RoleBinding associated with namespace "development"
clusterRole.bind(user, group); // creates a ClusterRoleBinding, not associated with any namespace
```

Future work:
- Add full `pod.grantRead(role)` and `Resources.fromTypes(ApiResource.POD).grantRead(role)` style APIs
- Add type checking or validation to prevent adding namespaced resources to ClusterRole or cluster-wide resources to Role's etc.
- Add more constants to `ApiResource` for granting permissions to [subresources](https://stackoverflow.com/questions/57872201/how-to-refer-to-all-subresources-in-a-role-definition)
- Add convenience APIs for adding permissions to the [default roles](https://kubernetes.io/docs/reference/access-authn-authz/rbac/#default-roles-and-role-bindings) using `aggregate-to-<default-role>` tags

BREAKING CHANGE: The interface `Resources` is now named `ContainerResources`.

Signed-off-by: Christopher Rybicki <rybickic@amazon.com>

Chriscbr committed:
feat: Role and ClusterRole L2s

Signed-off-by: Christopher Rybicki <rybickic@amazon.com>

Commit:
chore: self mutation

Signed-off-by: github-actions <github-actions@github.com>

Chriscbr committed:
add docstrings

Signed-off-by: Christopher Rybicki <rybickic@amazon.com>

iliapolo commented:
Suggestion for API:

```ts
const role = new Role(this, 'Role', { namespace: 'default'});
role.grantList(Grantee.fromObjects(deployment, pod));
role.grantList(Grantee.fromKind(Kind.Secrets, Kind.Deployments));
role.grantList({
  objects: [deployment, pod],
});
role.grantList({
  kinds: ['secrets', 'deployments'],
});

const binding = role.bind(user1, group1);
binding.addSubject(serviceAccount);
```

```ts
const role = new kplus.ClusterRole(chart, 'grafana-operator');
    role.addRule({
      apiGroups: [''],
      resources: ['events'],
      verbs: ['get', 'watch', 'list', 'create', 'delete', 'update', 'patch'],
    });
    role.addRule({
      apiGroups: ['integreatly.org'],
      resources: [
        'grafanadashboards',
        'grafanadatasources',
        'grafanadatasources/status',
      ],
      verbs: ['get', 'list', 'create', 'update', 'delete', 'deletecollection', 'watch'],
    });
```

Would be:

```ts
const role = new Role(this, 'Role');
role.grantGet(Kind.EVENTS);
role.grantWatch(Kind.EVENTS);
role.grantList(Kind.EVENTS);

role.grantGet(Kind.custom({ resources: 'grafanadashboards', 'grafanadatasources', apiGroup: 'integreatly.org'} 
```

Chriscbr committed:
more work

Signed-off-by: Christopher Rybicki <rybickic@amazon.com>

Chriscbr committed:
rename Kind to ResourceType

Signed-off-by: Christopher Rybicki <rybickic@amazon.com>

Chriscbr committed:
Merge branch 'k8s-22/main' into rybickic/roles

Signed-off-by: Christopher Rybicki <rybickic@amazon.com>

Chriscbr committed:
fix namespaces

Signed-off-by: Christopher Rybicki <rybickic@amazon.com>

Chriscbr commented on src/role.ts:
```diff
@@ -0,0 +1,317 @@
+  constructor(scope: Construct, id: string, props: RoleProps) {
+    super(scope, id);
+
+    if (props.metadata?.namespace && props.metadata.namespace !== props.namespace) {
+      throw new Error('If `metadata.namespace` is passed as an option, its value must match the value of `namespace`.');
+    }
+
+    const metadata: ApiObjectMetadata = {
+      ...props.metadata,
+      namespace: props.namespace,
+    };
```
I've exposed `namespace` as a top-level prop to make it clear that it's a required field and to make it easier to configure. But since namespace is also configurable through `metadata.namespace`, I had to add this error handling logic.

Anyway I'm open to changing this

iliapolo replied:
The fact `namespace` is required wasn't sitting right with me, so I tried to create a role without a namespace and apply - it worked...

I'm not sure why the docs explicitly say:

> A Role always sets permissions within a particular [namespace](https://kubernetes.io/docs/concepts/overview/working-with-objects/namespaces); when you create a Role, you have to specify the namespace it belongs in.

But I would actually skip this for now, its awkward and i'm not sure its worth it. Let's just stick to the normal namespace handling we have for all resources.

Chriscbr committed:
add RoleBinding and ClusterRoleBinding L2s

Signed-off-by: Christopher Rybicki <rybickic@amazon.com>

Chriscbr committed:
update docs

Signed-off-by: Christopher Rybicki <rybickic@amazon.com>

Chriscbr commented on src/grants.ts:
```diff
@@ -0,0 +1,54 @@
+    };
+  }
+
+  /**
+   * Combine multiple resource kinds into a single grantee that can be used in a
+   * Role or ClusterRole.
+   */
+  public static fromKinds(...types: ResourceType[]): IGrantee {
+    const apiGroups = normalizedArray(types
+      .map(type => type.apiGroups)
+      .flat(),
+    );
+    const resources = normalizedArray(types
+      .map(type => type.resources)
+      .filter((x): x is string[] => Boolean(x))
+      .flat(),
+    );
+    return { apiGroups, resources };
+  }
```
This isn't being used, since I designed `ResourceType` to implement `IGrantee` as is. Should I drop it?

Chriscbr commented on src/role.ts:
```diff
@@ -0,0 +1,360 @@
+    return undefinedIfEmpty(this._rules.map(rule => rule._toKube()));
+  }
+
+  /**
+   * Create a RoleBinding that binds the permissions in this Role
+   * to a list of subjects, that will only apply this role's namespace.
+   * @param subjects a list of subjects to bind to
+   */
+  public bind(...subjects: ISubject[]): RoleBinding {
+    const binding = new RoleBinding(this, 'RoleBinding', {
+      namespace: this.metadata.namespace!,
+      role: this,
+    });
+    binding.addSubjects(...subjects);
+    return binding;
+  }
```
Note: as implemented, you can't really call any of the bind methods (`bind`, `bindInNamespace`, `bindInCluster`) more than once on the same role because you'll get an error that there are two sibling constructs with the same name.

I don't think this is a blocker (since you can just create a RoleBinding directly and give it your own name) but thought it's worth calling out.

iliapolo replied:
I think this is ok for `bind` and `bindInCluster`, but not so much for `bindInNamespace`, see relevant comment.

Chriscbr committed:
rename and refactor

Signed-off-by: Christopher Rybicki <rybickic@amazon.com>

iliapolo commented:
@Chriscbr I think `ApiResource` is the eureka name we've been looking for! Good idea.  

Both `pods` and a single pod named `foo` represent an `ApiResource` (i.e one that can be managed via the API) - Its just that one has a `type`, while the other has a `name`. 

So, if we define a common interface:

```ts
export interface IApiResource {
  // mandatory, both pods and a single pod has this.
  readonly apiGroup: string;

  // optional - pods will have this
  readonly type?: string;

  // optional - a single will have this
  readonly name?: string;
}
```

Then, our `Resource` class actually already implements this interface.

```ts
export abstract class Resource extends Construct implements IResource, IApiResource {
```

If we now make `ApiResource` implement it:

```ts
export class ApiResource implements IApiResource {
  /**
   * API resource information for Binding.
   */
  public static readonly BINDING = new ApiResource({
    apiGroups: [''],
    resources: ['bindings'],
  });
  ....
  ....
}
```

We can change the allow methods to accept:

```ts
public allow(verbs: string[], ...resources: IApiResource[]): void {
```

And viola, we have the API we wanted:

```ts
role.allowRead(deployment, kplus.ApiResource.SECRETS, kplus.ApiResource.PODS);
```

Thoughts?

Chriscbr committed:
Merge branch 'k8s-22/main' into rybickic/roles

Signed-off-by: Christopher Rybicki <rybickic@amazon.com>

Chriscbr committed:
PR feedback

Signed-off-by: Christopher Rybicki <rybickic@amazon.com>

Chriscbr committed:
add docstrings

Signed-off-by: Christopher Rybicki <rybickic@amazon.com>

Chriscbr committed:
type to serviceType

Signed-off-by: Christopher Rybicki <rybickic@amazon.com>

Chriscbr committed:
rename type -> resourceType

Signed-off-by: Christopher Rybicki <rybickic@amazon.com>

Commit:
chore: self mutation

Signed-off-by: github-actions <github-actions@github.com>

Chriscbr commented:
Some more notes:
1) at the advice of Romain, I renamed `type` to `resourceType` to avoid naming conflicts in languages like Go.
2) I was re-reviewing the documentation and I think the "resources" field in rule is actually expected in all cases, as opposed to being an _alternative_ to "resourceNames". I've updated the PR based on this understanding, let me know if you have other thoughts

Ready for another round of reviews @iliapolo  :)

iliapolo commented on api-resources.txt:
```diff
@@ -0,0 +1,57 @@
+NAME                              SHORTNAMES   APIVERSION                             NAMESPACED   KIND                             VERBS
```
Maybe add a projen task that regenerates this file given access to a cluster? It will at least document the command we need to run.

iliapolo commented on projenrc/gen-api-resource.ts:
```diff
@@ -0,0 +1,284 @@
+
+/**
+ * Generate a source file for the `ApiResource` module, based on information about API
+ * resources generated by running `kubectl api-resources -o wide`.
```
Ah ok, command is here, but still, projen task could be nice :)

iliapolo commented on projenrc/gen-api-resource.ts:
```diff
@@ -0,0 +1,284 @@
+  ts.line('/**');
+  ts.line(' * The unique, namespace-global, name of an object inside the Kubernetes cluster');
+  ts.line(' */')
+  ts.line('readonly name?: string;')
```
For consistency:

```suggestion
  ts.line('readonly resourceName?: string;')
```

Also maybe add something like:

*if this is omitted, the ApiResource should represent all objects of the given type.*

iliapolo commented on projenrc/gen-api-resource.ts:
```diff
@@ -0,0 +1,284 @@
+    ts.line();
+  }
+
+  ts.open('public static custom(options: ApiResourceOptions): ApiResource {');
```
missing docstring

iliapolo commented on projenrc/gen-api-resource.ts:
```diff
@@ -0,0 +1,284 @@
+
+  for (const resource of resourceTypes) {
+    const typeName = normalizeTypeName(resource.kind);
+    const memberName = snakeCase(typeName.replace(/[^a-z0-9]/gi, '_')).split('_').filter(x => x).join('_').toUpperCase();
```
Can we do something to transform this to a plural form?
I'd rather we have `ApiResource.PODS` and not `ApiResource.POD`.

I see that the `NAME` field in the .txt file is already plural, so maybe:

```console
memberName = <what we have now> + toUpperCase(NAME - KIND)
```

?

Chriscbr replied:
Good idea. I had to do something slightly more involved because some English plurals are a bit funky (like "capacity" -> "capacities") but it does work 👍

iliapolo commented on src/config-map.ts:
```diff
@@ -52,14 +52,21 @@ export class ConfigMap extends Resource implements IConfigMap {
    */
   public static fromConfigMapName(name: string): IConfigMap {
-    return { name };
+    return {
```
You can reuse `k8s.KubeConfigMap.GVK`

iliapolo commented on src/container.ts:
```diff
@@ -520,7 +520,7 @@ export interface ContainerProps {
    * @see https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/
    */
-  readonly resources?: Resources;
+  readonly resources?: ContainerResources;
```
👍

iliapolo commented on src/pv.ts:
```diff
@@ -92,14 +92,21 @@ export class PersistentVolume extends Resource implements IPersistentVolume, ISt
    */
   public static fromPersistentVolumeName(volumeName: string): IPersistentVolume {
-    return { name: volumeName };
+    return {
```
`k8s.KubePersistentVolume.GVK`

iliapolo commented on src/pvc.ts:
```diff
@@ -85,14 +85,21 @@ export class PersistentVolumeClaim extends Resource implements IPersistentVolume
    */
   public static fromClaimName(claimName: string): IPersistentVolumeClaim {
-    return { name: claimName };
+    return {
```
`k8s.KubePersistentVolumeClaim.GVK`

iliapolo commented on src/secret.ts:
```diff
@@ -59,14 +59,21 @@ export class Secret extends Resource implements ISecret {
    */
   public static fromSecretName(name: string): ISecret {
-    return { name };
+    return {
```
`k8s.KubeSecret.GVK`

iliapolo commented on src/service-account.ts:
```diff
@@ -49,14 +42,21 @@ export class ServiceAccount extends Resource implements IServiceAccount {
    */
   public static fromServiceAccountName(name: string): IServiceAccount {
-    return { name: name };
+    return {
```
`k8s.KubeServiceAccount.GVK`

iliapolo commented on src/service-account.ts:
```diff
@@ -6,21 +6,14 @@ import { ISecret } from './secret';
-/**
- * Properties for initialization of `ServiceAccount`.
- */
-export interface ServiceAccountProps extends ResourceProps {
```
Thank you

iliapolo commented on src/service.ts:
```diff
@@ -165,7 +165,7 @@ export class Service extends Resource {
    * Determines how the Service is exposed.
    */
-  public readonly type: ServiceType;
+  public readonly serviceType: ServiceType;
```
I think this is leftover from when `IApiResource` had a `type` property? We don't have to change this now.

iliapolo commented on src/role.ts:
```diff
@@ -0,0 +1,380 @@
+export interface RoleProps extends RoleCommonProps {
+  /**
+   * The namespace of the resources this role's permissions should apply to
+   * (required).
```
```suggestion
   * The namespace this role is created in. Permissions added to this role must refer to resources in the same namespace
```

iliapolo commented on src/role.ts:
```diff
@@ -0,0 +1,380 @@
+}
+
+/**
+ * Options for `Role`.
```
```suggestion
 * Properties for `Role`.
```

iliapolo commented on test/role.test.ts:
```diff
@@ -0,0 +1,406 @@
+    const role = new kplus.Role(chart, 'pod-reader', {
+      namespace: 'default',
+    });
+    role.addRule({
```
You should also assert the api of the returned `PolicyRule` instance.

iliapolo commented on src/role.ts:
```diff
@@ -0,0 +1,380 @@
+import { filterUndefined, undefinedIfEmpty } from './utils';
+
+export interface RoleCommonProps extends ResourceProps {
+  readonly rules?: PolicyRuleProps[];
```
docstring

iliapolo commented on src/role.ts:
```diff
@@ -0,0 +1,380 @@
+import { ClusterRoleBinding, ISubject, RoleBinding } from './role-binding';
+import { filterUndefined, undefinedIfEmpty } from './utils';
+
+export interface RoleCommonProps extends ResourceProps {
```
docstring

iliapolo commented on src/role.ts:
```diff
@@ -0,0 +1,380 @@
+  readonly rules?: PolicyRuleProps[];
+}
+
+export abstract class RoleBase extends Resource implements IResource {
```
docstring

iliapolo commented on src/role.ts:
```diff
@@ -0,0 +1,380 @@
+   * to API resources (such as "pods" or "secrets") or non-resource URL paths
+   * (such as "/api"),  but not both.
+   */
+  readonly nonResourceUrLs?: string[];
```
The docs state:

> this field is only applicable for ClusterRoles referenced from a ClusterRoleBinding

This means this property shouldn't be available when defining a `Role`, only a `ClusterRole`. 
I would create a `RolePolicyRule` and a `ClusterRolePolicyRule` that extends it. 

Also

> Rules can either apply to API resources (such as "pods" or "secrets") or non-resource URL paths (such as "/api"),  but not both.

This means `ClusterRole` will have to validate only one of them is used. (A `Role` won't have to validate since this option won't even be available to it).

iliapolo commented on src/role.ts:
```diff
@@ -0,0 +1,380 @@
+   * to API resources (such as "pods" or "secrets") or non-resource URL paths
+   * (such as "/api"),  but not both.
+   */
+  readonly nonResourceUrLs?: string[];
```
The docs state:

> this field is only applicable for ClusterRoles referenced from a ClusterRoleBinding

This means this property shouldn't be available when defining a `Role`, only a `ClusterRole`. 
I would create a `RolePolicyRule` and a `ClusterRolePolicyRule` that extends it. 

Also

> Rules can either apply to API resources (such as "pods" or "secrets") or non-resource URL paths (such as "/api"),  but not both.

This means `ClusterRole` will have to validate only one of them is used. (A `Role` won't have to validate since this option won't even be available to it).

Chriscbr replied:
Interesting, this is a good catch.

I was thinking about this for a bit, and I realized if we say a `RolePolicyRule` is a rule about API resources, and a `ClusterRolePolicyRule` is a rule about API resources OR non-resources, it actually makes more sense for the subtyping relation to go the other way around -- that is, `RolePolicyRule` is a subtype (a special case of) `ClusterRolePolicyRule`!

To make the subtyping relationship clearer I've renamed them `PolicyRule` and `ResourcePolicyRule` respectively (see updated PR). Thoughts?

iliapolo commented on src/role.ts:
```diff
@@ -0,0 +1,380 @@
+ * ClusterRole is a cluster level, logical grouping of PolicyRules that can be
+ * referenced as a unit by a RoleBinding or ClusterRoleBinding.
+ */
+export class ClusterRole extends RoleBase implements IClusterRole {
```
I think this deserves its own file, since its a distinct resource.

Chriscbr replied:
I'd rather leave this as is since to get some inheritance behavior working I need `Role` and `ClusterRole` to both have access to the un-exported `RoleBase` class (otherwise JSII throws errors preventing me from giving `Role.addRule(rule: ResourcePolicyRuleProps)` a different signature from `RoleBase.addRule(rule: PolicyRuleProps)`).

iliapolo commented on test/role.test.ts:
```diff
@@ -0,0 +1,406 @@
+    ]));
+
+  });
+
```
I think a mixed test is appropriate as well, e.g:

```ts
role.allowRead(deployment, kplus.ApiResource.POD);
```

iliapolo commented on test/role-binding.test.ts:
```diff
@@ -0,0 +1,186 @@
+  // WHEN
+  const user = new kplus.User({ name: 'alice@example.com' });
+  const group = new kplus.Group({ name: 'frontend-admins' });
+  role.bind(user, group);
```
Add assertions on the `RoleBinding` api that is returned.

iliapolo commented on src/role.ts:
```diff
@@ -0,0 +1,380 @@
+   * @param subjects a list of subjects to bind to
+   */
+  public bindInNamespace(namespace: string, ...subjects: ISubject[]): RoleBinding {
+    const binding = new RoleBinding(this, 'RoleBinding', {
```
This is problematic. From the docs:

> This kind of reference lets you define a set of common roles across your cluster, then reuse them within multiple namespaces.

i.e, calling this method multiple times is actually a common use case. Fortunately there's an easy fix:

```suggestion
    const binding = new RoleBinding(this, `RoleBinding${namespace}`, {
```

iliapolo commented on src/role.ts:
```diff
@@ -0,0 +1,380 @@
+   * ClusterRole to a list of subjects, without namespace restrictions.
+   * @param subjects a list of subjects to bind to
+   */
+  public bindInCluster(...subjects: ISubject[]): ClusterRoleBinding {
```
I think we can use `bind` here. 

```suggestion
  public bind(...subjects: ISubject[]): ClusterRoleBinding {
```

This way its consistent with the `bind` API in `Role`. So we have:

- `role.bind -> RoleBinding`
- `clusterRole.bind -> ClusterRoleBinding`
- `clusterRole.bindInNamespace` -> `RoleBinding`.

iliapolo commented on src/role-binding.ts:
```diff
@@ -0,0 +1,247 @@
+ * A ClusterRoleBinding grants permissions cluster-wide to a user or
+ * set of users.
+ */
+export class ClusterRoleBinding extends Resource {
```
I think this deserves its own file, since its a distinct resource.

Chriscbr replied:
To align with `roles.ts` I'll leave this as is

iliapolo commented on src/role.ts:
```diff
@@ -0,0 +1,380 @@
+    });
+  }
+
+  private synthesizeRules(): any {
```
```suggestion
  private synthesizeRules(): k8s.PolicyRule[] {
```

iliapolo commented on src/role.ts:
```diff
@@ -0,0 +1,380 @@
+ * about who the rule applies to or which namespace the rule applies to.
+ */
+export class PolicyRule {
+  constructor(public readonly props: PolicyRuleProps) {}
```
We normally don't just expose `props` as a public property. Flatten these out to the individual properties it contains - makes for a simpler API for consumers.

iliapolo commented on src/role-binding.ts:
```diff
@@ -0,0 +1,247 @@
+}
+
+/**
+ * Options for `RoleBinding`.
```
```suggestion
 * Properties for `RoleBinding`.
```

iliapolo commented on src/role-binding.ts:
```diff
@@ -0,0 +1,247 @@
+  /**
+   * Namespace of the resources the RoleBinding should apply to.
+   */
+  readonly namespace: string;
```
I dont think this should be required. The default should be to use the role's namespace (even if its undefined). 

I also wonder what happens if the `namespace` is different than that of the role's, is that allowed? can you verify by trying to apply such a manifest?

iliapolo commented on src/role-binding.ts:
```diff
@@ -0,0 +1,247 @@
+/**
+ * Options for `User`.
+ */
+export interface GroupProps {
```
```suggestion
export interface UserProps {
```

Chriscbr replied:
Oops

iliapolo commented on src/role-binding.ts:
```diff
@@ -0,0 +1,247 @@
+}
+
+/**
+ * Options for `User`.
```
```suggestion
 * Properties for `User`.
```

iliapolo commented on src/role-binding.ts:
```diff
@@ -0,0 +1,247 @@
+  /**
+   * The namespace that the group applies to.
+   */
+  readonly namespace?: string;
```
Something doesn't add up, the schema docs for this property say:

> If the object kind is non-namespace, such as "User" or "Group", and this value is not empty the Authorizer should report an error.

Are you sure a group is a namespaced resource? I think this property is only relevant for a `ServiceAccount` subject.

Chriscbr replied:
Yep I believe you're correct

iliapolo commented on src/role-binding.ts:
```diff
@@ -0,0 +1,247 @@
+ */
+export interface GroupProps {
+  /**
+   * The name of the user.
```
```suggestion
   * The name of the group.
```

iliapolo commented on src/role-binding.ts:
```diff
@@ -0,0 +1,247 @@
+}
+
+/**
+ * Options for `Group`.
```
```suggestion
 * Properties for `Group`.
```

iliapolo commented on src/role-binding.ts:
```diff
@@ -0,0 +1,247 @@
+ * applies to.  This can either hold a direct API object reference, or a value
+ * for non-objects such as user and group names.
+ */
+export interface ISubject {
```
I expected to see `ServiceAccount` implement this interface. Is this something you didn't get to?

Chriscbr replied:
Ah I had a test for this, but it was only working because of structural typing. Fixed

iliapolo commented on src/role.ts:
```diff
@@ -0,0 +1,380 @@
+   * Add the rules of the argument ClusterRole into this ClusterRole.
+   * @param role
+   */
+  public aggregateFrom(role: ClusterRole): void {
```
I'm not sure about this API. Does an aggregation only makes sense with another managed role? What if I just need to add an aggregation label, pointing to an existing `ClusterRole`?

How about:

```ts
// aggregate rules from roles matching this label selector
public aggregate(key: string, value: string): void;

// combine rules from the given role by applying the appropriate aggregation labels
public combine(role: ClusterRole): void;
```

Chriscbr replied:
Good idea 👍

iliapolo reviewed (changes requested):
Sweet! This is a big one.

Chriscbr committed:
Merge branch 'k8s-22/main' into rybickic/roles

Signed-off-by: Christopher Rybicki <rybickic@amazon.com>

Chriscbr committed:
PR feedback

Signed-off-by: Christopher Rybicki <rybickic@amazon.com>

Chriscbr committed:
pluralize resources

Signed-off-by: Christopher Rybicki <rybickic@amazon.com>

Chriscbr committed:
use GVKs

Signed-off-by: Christopher Rybicki <rybickic@amazon.com>

Chriscbr committed:
revert type -> serviceType

Signed-off-by: Christopher Rybicki <rybickic@amazon.com>

Chriscbr committed:
fix docstrings

Signed-off-by: Christopher Rybicki <rybickic@amazon.com>

Chriscbr committed:
update tests

Signed-off-by: Christopher Rybicki <rybickic@amazon.com>

Chriscbr committed:
update docstrings

Signed-off-by: Christopher Rybicki <rybickic@amazon.com>

Chriscbr commented:
Still WIP

Chriscbr committed:
separate PolicyRule into two types

Signed-off-by: Christopher Rybicki <rybickic@amazon.com>

Chriscbr committed:
make namespace not required in Role

Signed-off-by: Christopher Rybicki <rybickic@amazon.com>

Chriscbr committed:
implement ISubject on ServiceAccount

Signed-off-by: Christopher Rybicki <rybickic@amazon.com>

Chriscbr committed:
can bind to namespace multiple times

Signed-off-by: Christopher Rybicki <rybickic@amazon.com>

Chriscbr committed:
bindInCluster -> bind

Signed-off-by: Christopher Rybicki <rybickic@amazon.com>

Chriscbr committed:
flatten out PolicyRule properties

Signed-off-by: Christopher Rybicki <rybickic@amazon.com>

Chriscbr committed:
add tests and remove faulty rule merging

Signed-off-by: Christopher Rybicki <rybickic@amazon.com>

Chriscbr committed:
update RoleBinding tests

Signed-off-by: Christopher Rybicki <rybickic@amazon.com>

Chriscbr committed:
refactor aggregation API

Signed-off-by: Christopher Rybicki <rybickic@amazon.com>

Chriscbr committed:
improve comments

Signed-off-by: Christopher Rybicki <rybickic@amazon.com>

Chriscbr committed:
Merge branch 'k8s-22/main' into rybickic/roles

Signed-off-by: Christopher Rybicki <rybickic@amazon.com>

Chriscbr commented:
Thanks for all of the feedback @iliapolo ! This should be ready for another review 🙂

iliapolo committed:
Merge branch 'k8s-22/main' into rybickic/roles

Signed-off-by: Eli Polonsky <epolon@amazon.com>

iliapolo committed:
Merge branch 'k8s-22/main' into rybickic/roles

Signed-off-by: Eli Polonsky <epolon@amazon.com>

Commit:
chore: self mutation

Signed-off-by: github-actions <github-actions@github.com>

iliapolo committed:
chore: roles refactor (#678)

@Chriscbr Following our conversation, I think we need to take a somewhat different approach. 

We've been trying to optimize for code maintenance by creating a common `RoleBase` that gives us some common functionality between a `Role` and a `ClusterRole`. However, we should be optimizing for user experience. 

A `Role` can only define rules for resources, where as a `ClusterRole` can define rules both for resources and non-resources. Trying to express this inherent difference, while still keeping a common base class, is I think what gets us in trouble. The solution we came up with is to have a common `addResourceRule` method in the base class, and have `ClusterRole` add an `addNonResourceRule` method that will only apply to it. 

This will result in the following API:

```ts
const role = new kplus.Role(...);
role.addResourceRule(...);
role.allow(...);

const clusterRole = new kplus.Role(...);
clusterRole.addResourceRule(...);
clusterRole.addNonResourceRule(...);
clusterRole.allow(...);
```

There are several problems with this:

- Naming inconsistency (`addResourceRule` vs `allow` - even though both refer to resource rules). Note that if we change `addResourceRule -> addRule`, it won't be clear why `ClusterRole` has `addRule` and `addNonResourceRule`.
- The `add` methods don't offer any capabilities that the `allow` method don't. It creates a situation where we have two API's offering the same thing - which is confusing. I'm not sure the `add` methods have merit in higher level API's, the whole point is to remove this bulky k8s structure. If users want to use the underlying structure instead, they can always escape hatch. 

If we focus on optimizing user experience for the common use-case, I think the API should look like this: 

```ts
const role = new kplus.Role(...);
// allow reading a specific pod and all secrets
role.allowGet(pod, ApiResource.SECRETS);

const clusterRole = new kplus.ClusterRole(...);
// allow reading a specific pod, all secrets, and the /healthz endpoint
clusterRole.allowGet(pod, ApiResource.SECRETS, ApiEndpoint.of('/healthz'));
```

Note that a `Role` can be given only api resources, but a `ClusterRole` can be given non api resources as well, all with the same API. 

This PR adds an `ApiEndpoint` (this name is also used in k8s docs: *non-resource endpoints (like /healthz)*) interface that allows `ClusterRole` to accept everything it can. It also removes the `add` methods in favor of the existing `allow` ones. 

Note that the base was class was completely removed, which of course resulted in a some duplication, but I think its worth it for the flexibility it gives. 

#### Resources

- https://octopus.com/blog/k8s-rbac-roles-and-bindings
- https://kubernetes.io/docs/reference/access-authn-authz/rbac/

iliapolo committed:
Merge branch 'k8s-22/main' into rybickic/roles

Signed-off-by: Eli Polonsky <epolon@amazon.com>

iliapolo committed:
Merge branch 'k8s-22/main' into rybickic/roles

Signed-off-by: Eli Polonsky <epolon@amazon.com>

iliapolo committed:
projen

Signed-off-by: Eli Polonsky <epolon@amazon.com>

cdk8s-automation commented:
## 💔 All backports failed

| Status | Branch | Result |
|:------:|:------:|:------|
|❌|k8s-21/main|**Backport failed because of merge conflicts**|
|❌|k8s-20/main|**Backport failed because of merge conflicts**|

### Manual backport
To create the backport manually run:
```
backport --pr 432
```

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

<!--- Backport version: 8.4.0 -->

iliapolo commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|k8s-21/main|[<img src="https://img.shields.io/github/pulls/detail/state/cdk8s-team/cdk8s-plus/685">](https://github.com/cdk8s-team/cdk8s-plus/pull/685)|

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

<!--- Backport version: 8.4.0 -->

iliapolo commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|k8s-20/main|[<img src="https://img.shields.io/github/pulls/detail/state/cdk8s-team/cdk8s-plus/686">](https://github.com/cdk8s-team/cdk8s-plus/pull/686)|

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

<!--- Backport version: 8.4.0 -->
